### ISSUE TITLE
Editor for notebook cells opens, not notebook editor

### ISSUE THREAD

**Issue by jrieken** [microsoft/vscode#123270]
* Have an extension that uses the `showTextDocument`-API for a cell document
* :bug: instead of the notebook editor, a text editor is shown

**Comment by jrieken** [comment]
@lramos15 It seems that you have done on purpose with this line: https://github.com/microsoft/vscode/blob/413963c489fafa5163b5d6513731c7953de07fb3/src/vs/workbench/api/browser/mainThreadEditors.ts#L145

I wonder why I would object against this.

**Comment by jrieken** [comment]
I see how you just migrated that into the new world, it was already wrong before. I will change it and see what happens ;-)

**Comment by jrieken** [comment]
I do understand... This is because of https://github.com/microsoft/vscode/issues/96397 and it has its merit

@lramos15 I wonder if you can give me another options, like `DISABLED` unless there is an exclusive handler. Or change the meaning of disables to that, ignore when there is an exclusive handler

**Comment by cedric05** [comment]
will there be built in xml/ mimetype=`application/xml` renderer. currently there isn't . how to know list of vscode built-in renderers & can I override builtin renderers?

**Comment by jrieken** [comment]
@cedric05 Unrelated issue?

**Comment by lramos15** [comment]
@jrieken I know we discussed this but the more I think about it the more I get confused about a call to `showTextDocument` opening anything but a text editor. Wouldn't it make more sense for extension authors to use `vscode.open`. I feel like we will further run into issues like the git one even if we only allow exclusive internal providers

**Comment by lramos15** [comment]
I've implemented another override option that is `EXCLUSIVE_ONLY`, but when adding that to https://github.com/microsoft/vscode/blob/main/src/vs/workbench/api/browser/mainThreadEditors.ts#L145 I see integration test failures. When I try to debug the integration tests it just keeps crashing my OSS build. If you have any suggestions let me know.

**Comment by rebornix** [comment]
@lramos15 I was going through skipped tests and found this one. Tried to adopt `EditorOverride.EXCLUSIVE_ONLY` but the test is still failing (opening a text editor other than a notebook editor). Let me know if you can reproduce without crash, if that still happens, we can take a look together.

**Comment by rebornix** [comment]
@lramos15 and I looked into this today and tried to fix it, there are multiple things that need to be changed/fixed to get it working.

Foremost, `this._editorService.openEditor` can return a composite editor, so trying to match it with `textEditor` won't work. We need to get its `acitveCodeEditor` and compare if the editor pane is a composite editor. 

With this change and changing `override: EditorOverride.DISABLED` to `override: EditorOverride.EXCLUSIVE_ONLY` will break [window api tests](https://insiders.vscode.dev/github/microsoft/vscode-distro/blob/336373dfc5088b5a679d3353b6c81e697526cae7/extensions/vscode-api-tests/src/singlefolder-tests/window.test.ts#L45)  (normal `showTextDocument` tests). We even changed it to `override: 'default'` which also failed the window api tests.

Here is what happens, when we call `MainthreadTextEditors.$tryShowTextDocument`

* When `ITextEditorOptions.override` is set to `Disabled`
  * Before `_editorService.openEditor` resolves
    * `ICodeEditor` is created
    * `IMainThreadEditorLocator` receives the event, but ignores it as the editor has no model yet
    * `NativeTextFileEditor.setInput` resolves, which sets a valid model on the editor
    * `IMainThreadEditorLocator` receives the event, and broadcasts the editor/document addition
  * When `_editorService.openEditor` resolves, we now have a valid editor with model
    * We can find this editor id from `IMainThreadEditorLocator.findTextEditorIdFor` 
* When `ITextEditorOptions.override` is set to other value, say `default` (forcing it to use text editor)
  * Before `_editorService.openEditor` resolves
    * `ICodeEditor` is created
    * `IMainThreadEditorLocator` receives the event, but ignores it as the editor has no model yet
  * Then `_editorService.openEditor` resolves
    * We check `IMainThreadEditorLocator.findTextEditorIdFor`, which can't find any matching editor (since there is no valid editor with model yet)
	* Return `undefined`
	* ExtHost throws error as `MainthreadTextEditors.$tryShowTextDocument` doesn't give back a valid editor it
  * `NativeTextFileEditor.setInput` resolves, which sets a valid model on the editor
    * `IMainThreadEditorLocator` receives the event, and broadcasts the editor/document addition
    * But it's already too late

Above is easy to reproduce by changing https://github.com/microsoft/vscode/blob/main/src/vs/workbench/api/browser/mainThreadEditors.ts#L128 to `default` or `EditorOverride.EXCLUSIVE_ONLY` and run integration tests. @bpasero we would appreciate your help if you could give us pointers to how to get this resolved.

**Comment by lramos15** [comment]
To add, there should be no difference between `EditorOverride.DISABLED` and `'default'` as they both call upon the `textEditorService.createTextEditor`. My guess is this that the resolver is slightly slower to create the input than just calling the `textEditorService` directly from the `editorService` causing this behavior to surface

**Comment by bpasero** [comment]
Let us wait for Jo to be back and setup a meeting, I can understand how the editor service works but I have no stakes in how editors travel to the extension host. 

It is true however that there is a difference between using the editor resolver or not: if the editor resolver is involved, there is a `Promise` and otherwise not:

https://github.com/microsoft/vscode/blob/cebad239e88e15b3738844e77cf26955d6b4d1a7/src/vs/workbench/services/editor/browser/editorService.ts#L520-L522

However since everything in the editor service is already promise based and imho the only reason we have promises in the resolver is to support the case of picking an editor by the user, execution should really not differ at all, unless there is some side effect in the editor resolver that could trigger this?

**Comment by bpasero** [comment]
When debugging into the resolver, I see no evidence of code that executes `async`, everything seems to be `sync`. However, maybe V8 yields back after calling N methods that are `async` for some reason or I am missing something.

However, what I do see is that the following line is hit 3 times for each `showTextDocument` call before returning:

https://github.com/microsoft/vscode/blob/cebad239e88e15b3738844e77cf26955d6b4d1a7/src/vs/workbench/services/editor/browser/editorService.ts#L505

Whereas previously this line is hit in sequence:

https://github.com/microsoft/vscode/blob/cebad239e88e15b3738844e77cf26955d6b4d1a7/src/vs/workbench/services/editor/browser/editorService.ts#L521

So it is possible that we now resolve the same editor 3 times in parallel when previously we would resolve it in sequence and that causes the test issue.

In other words, when an extension calls `showTextDocument` in parallel promises, the execution flow is now different from before.

**Comment by bpasero** [comment]
Oh, AlexD was nice enough to explain to me that any `async` statement will cause the engine to yield to the next runner and only execute at the end of the loop before the next loop.

**Comment by rebornix** [comment]
@bpasero great analysis!

Writing tests in **sequence** fashion will pass 

```ts
		const doc = await workspace.openTextDocument(join(workspace.rootPath || '', './far.js'));
		const editor = await window.showTextDocument(doc, ViewColumn.One)
		assert.strictEqual(editor.viewColumn, ViewColumn.One);
		const editor_1 = await window.showTextDocument(doc, ViewColumn.Two);
		assert.strictEqual(editor_1.viewColumn, ViewColumn.Two);
		const editor_2 = await window.showTextDocument(doc, ViewColumn.Three);
		assert.strictEqual(editor_2.viewColumn, ViewColumn.Three);
```

though I think the tests are intentionally calling show documents without `await` and still expecting them to be created with right view column (being handled in the same sequence as how `window.showTextDocument` is called) 

 https://github.com/microsoft/vscode/blob/d9c6f61294f6e21a258673ee19b3e7aef01c5cfb/extensions/vscode-api-tests/src/singlefolder-tests/window.test.ts#L43-L51

**Comment by bpasero** [comment]
The problem is with the test expecting to open 3 groups at the same time by asking to open 3 editors in 3 groups, but we have a condition that will normalize the group when we cannot find it:

https://github.com/microsoft/vscode/blob/07e45c5a71b338bac1815417f4c69d29cd643d05/src/vs/workbench/services/editor/common/editorGroupColumn.ts#L18-L23

Without the `async` execution, we manage to create the group and have our first editor be in there, so that we correctly assign `SIDE_GROUP` for the second and third editor open request.

With `async` execution, they all end up in the `ACTIVE_GROUP` because the groups have not been created yet. 

Even if we take out this line:

https://github.com/microsoft/vscode/blob/07e45c5a71b338bac1815417f4c69d29cd643d05/src/vs/workbench/services/editor/common/editorGroupColumn.ts#L21

It would not work with 3 editors, only 2 because the last 2 requests would end up in the same side group.

I currently do not really have a good idea how to solve that other than maybe fundamentally change how we map view columns to editor groups, e.g. by creating editor groups eagerly even before opening an editor within. Not really sure about the consequences tbh.

**Comment by rebornix** [comment]
@bpasero I just found that the `showNotebookDocument` ran into a similar problem 

![new-notebook-untitled](https://user-images.githubusercontent.com/876920/183224462-12bd80d0-2498-48d0-bd40-2870a726ad94.gif)


I was running OSS on Windows with Jupyter extension installed. Then run command "New Jupyter Notebook" twice quickly. `this._editorService.openEditor` will return `undefined` for the first untitled notebook document. In notebook, we throw error if it's `undefined` while in text document, we return `undefined` to extension host.

https://github.com/microsoft/vscode/blob/cb79d015557d546d0d48cea97d7fc3ac0d3dc590/src/vs/workbench/api/browser/mainThreadNotebookEditors.ts#L130-L137

Note that this is creating editors in the same group but not sure how it fails internally in `this._editorService.openEditor`.

**Comment by bpasero** [comment]
There was a relatively recent change (6 months?) to return `undefined` if opening fails. Previously we would just return the editor that is active but that was very misleading. How upstream components deal with `undefined` is up to them.

**Comment by bpasero** [comment]
@rebornix @lramos15 not sure if my PR also fixed this issue or is there more to it? Then we should reopen.

**Comment by rebornix** [comment]
A new issue just came in https://github.com/microsoft/vscode/issues/158412, not sure if it's related to this change, will follow up.

**Comment by bpasero** [comment]
I am not sure where we are on the discussion, but my change did this:

https://github.com/microsoft/vscode/blob/f5c96e2db885f43fe1adb4a6ebbdfecf095430a8/src/vs/workbench/api/browser/mainThreadEditors.ts#L130

Where I think @lramos15 indicated in a discussion that `EditorResolution.EXCLUSIVE_ONLY` should be used?
